src/explain.rs

```diff
@@ -0,0 +1,806 @@
+                !self.explainfind[usize::from(node)].is_rewrite_forward;
+            self.explainfind[usize::from(next)].next = node;
+        }
+    }
+
+    pub(crate) fn union(
+        &mut self,
+        node1: Id,
+        node2: Id,
+        canon_id1: Id,
+        canon_id2: Id,
+        justification_maybe: Option<Justification>,
+    ) -> Id {
+        assert_ne!(self.find(node1), self.find(node2));
+        if let Some(justification) = justification_maybe {
+            self.make_leader(node1);
+            self.explainfind[usize::from(node1)].next = node2;
+            self.explainfind[usize::from(node1)].justification = justification;
+            self.explainfind[usize::from(node1)].is_rewrite_forward = true;
+            self.unionfind.union_roots(canon_id1, canon_id2)
+        } else {
+            // Act like a normal union find when the rule is not provided.
+            self.unionfind.union_roots(canon_id1, canon_id2)
+        }
+    }
+
+    pub fn explain_matches(
+        &mut self,
+        left: &RecExpr<L>,
+        right: &PatternAst<L>,
+        subst: &Subst,
+    ) -> Explanation<L> {
+        let left_added = self.add_expr(left);
+        let right_added = self.add_match(None, &right, &subst);
+        Explanation::new(self.explain_enodes(left_added, right_added))
+    }
+
+    pub fn explain_equivalence(&mut self, left: &RecExpr<L>, right: &RecExpr<L>) -> Explanation<L> {
+        let left_added = self.add_expr(left);
+        let right_added = self.add_expr(right);
+        Explanation::new(self.explain_enodes(left_added, right_added))
+    }
+
+    fn common_ancestor(&self, mut left: Id, mut right: Id) -> Id {
+        let mut seen_left: HashSet<Id> = Default::default();
+        let mut seen_right: HashSet<Id> = Default::default();
+        loop {
+            seen_left.insert(left);
+            if seen_right.contains(&left) {
+                return left;
+            }
+
+            seen_right.insert(right);
+            if seen_left.contains(&right) {
+                return right;
+            }
+
+            let next_left = self.explainfind[usize::from(left)].next;
+            let next_right = self.explainfind[usize::from(right)].next;
+            assert!(next_left != left || next_right != right);
+            left = next_left;
+            right = next_right;
+        }
+    }
+
+    fn get_nodes<'a>(&'a self, mut node: Id, ancestor: Id) -> Vec<&'a ExplainNode<L>> {
+        if node == ancestor {
+            return vec![];
+        }
+
+        let mut nodes = vec![];
+        loop {
+            let next = self.explainfind[usize::from(node)].next;
+            nodes.push(&self.explainfind[usize::from(node)]);
+            if next == ancestor {
+                return nodes;
+            }
+            assert!(next != node);
+            node = next;
+        }
+    }
+
+    fn explain_enodes(&self, left: Id, right: Id) -> ExplanationTrees<L> {
```

Comment:
This function is the whole proof-production algorithm. The expression corresponding to a node is guaranteed to match any rules it is connected to, so rewrites are easy to prove.